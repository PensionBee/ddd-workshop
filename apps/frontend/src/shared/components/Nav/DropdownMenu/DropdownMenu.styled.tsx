import styled from "styled-components";

export const StyledContentList = styled("ul")`
  display: grid;
  padding: 16px 0px;
  margin: 0px;
  column-gap: 10px;
  list-style: none;
  @media only screen and (min-width: 600px) {
    width: 250px;
  }
`;
