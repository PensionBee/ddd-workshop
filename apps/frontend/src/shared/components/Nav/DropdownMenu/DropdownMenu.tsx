import { useFocusOutside } from "@shared/hooks/useFocusOutside/useFocusOutside";
import { useEffect, useState } from "react";
import { useRef } from "react";
import { useArrowNavigation } from "../hooks/useArrowNavigation/useArrowNavigation";
import type { NavMenuItem } from "../Nav.types";
import { StyledContentList } from "./DropdownMenu.styled";
import { DropdownMenuSubList } from "./DropdownMenuSubList/DropdownMenuSubList";

export type DropdownMenuProps = {
  subList: NavMenuItem["subList"];
  handleCloseDropdown?: () => void;
};

export const DropdownMenu: React.FC<DropdownMenuProps> = ({
  subList,
  handleCloseDropdown,
  ...props
}) => {
  const [isDropdownOpen, setIsDropdownOpen] = useState(false);
  const dropdownRef = useRef<HTMLUListElement | null>(null);

  useFocusOutside(dropdownRef, () => handleCloseDropdown?.());

  useArrowNavigation(dropdownRef, {
    direction: "vertical",
    disabled: isDropdownOpen,
    onExitTop: handleCloseDropdown,
  });

  useEffect(() => {
    const firstChild = dropdownRef.current?.querySelector(
      "a,button"
    ) as HTMLElement | null;
    firstChild?.focus();
  }, []);

  return (
    <StyledContentList
      data-rc="DropdownMenu"
      ref={dropdownRef}
      data-state={isDropdownOpen ? "open" : "closed"}
      {...props}
    >
      <DropdownMenuSubList
        subList={subList}
        setIsDropdownOpen={setIsDropdownOpen}
      />
    </StyledContentList>
  );
};
