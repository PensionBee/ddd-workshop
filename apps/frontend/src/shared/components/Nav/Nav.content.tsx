import type { NavMenuItem } from "./Nav.types";

// Example content used in tests and stories

export const VISIBLE_LINK: NavMenuItem = {
  children: "Visible Item",
  href: "/",
};
export const HIDDEN_LINK: NavMenuItem = {
  children: "Hidden Item",
  href: "/",
  hidden: true,
};
export const SUB_LIST_ITEM: NavMenuItem = {
  children: "SubList Item",
  subList: [VISIBLE_LINK],
};
export const HIDDEN_SUB_LIST_ITEM: NavMenuItem = {
  children: "Hidden SubList Item",
  subList: [VISIBLE_LINK],
  hidden: true,
};

export const LIST = [VISIBLE_LINK] satisfies NavMenuItem[];
export const LIST_WITH_HIDDEN_LINK = [HIDDEN_LINK] satisfies NavMenuItem[];
export const LIST_WITH_SUB_LIST_ITEM = [SUB_LIST_ITEM] satisfies NavMenuItem[];
export const LIST_WITH_MIXED_ITEMS = [
  VISIBLE_LINK,
  HIDDEN_LINK,
  SUB_LIST_ITEM,
  HIDDEN_SUB_LIST_ITEM,
] satisfies NavMenuItem[];
