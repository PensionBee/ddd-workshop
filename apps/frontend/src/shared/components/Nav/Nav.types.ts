import type { LargeSystemIcon } from "@shared/components/Svg";

export type NavMenuItem = {
  children: string;
  href?: string;
  subList?: NavMenuItem[];
  hidden?: boolean;
  icon?: React.ComponentProps<typeof LargeSystemIcon>["icon"];
};
