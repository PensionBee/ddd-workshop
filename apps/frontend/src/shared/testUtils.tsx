import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import type { Breakpoints } from "./styling/breakpoints";
import { BREAKPOINTS } from "./styling/breakpoints";
import { theme } from "./styling/theme";
import { vi } from "vitest"

type ReactTestingLibraryProviderProps = {
  children: React.ReactNode;
};

// eslint-disable-next-line react-refresh/only-export-components
const ReactTestingLibraryProviders: React.FC<
  ReactTestingLibraryProviderProps
> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const customRender = (ui: React.ReactElement, options?: any) => {
  return render(ui, { wrapper: ReactTestingLibraryProviders, ...options });
};

// re-export everything
// eslint-disable-next-line react-refresh/only-export-components
export * from "@testing-library/react";

// override render method
export { customRender as render };

export const sleep = (delay: number) => {
  return new Promise((resolve) => setTimeout(resolve, delay));
};

export const mockMediaMatches = (breakpoint: Breakpoints) => {
  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: (query: string) => {
      const breakpointValue = BREAKPOINTS[breakpoint];
      const queryValue = parseInt(query.replace(/^\D+/g, ""));
      const breakpointValues = [...Object.values(BREAKPOINTS)];
      if (!breakpointValues.includes(queryValue)) {
        throw new Error(
          `Invalid breakpoint value: ${queryValue}, this likely means there has been a change to the way that breakpoints are setup.`
        );
      }
      return {
        matches: queryValue <= breakpointValue,
        addEventListener: vi.fn(),
        removeEventListener: vi.fn(),
      };
    },
  });
};

export const routerRender = (ui: React.ReactNode, pathname = "/") => {
  return customRender(
    <MemoryRouter initialEntries={[{ pathname }]}>{ui}</MemoryRouter>
  );
};
