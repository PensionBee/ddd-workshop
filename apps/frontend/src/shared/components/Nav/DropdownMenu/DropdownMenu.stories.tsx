import type { ComponentStory, ComponentMeta } from "@storybook/react";
import { LIST_WITH_MIXED_ITEMS } from "../Nav.content";
import type { DropdownMenuProps } from "./DropdownMenu";
import { DropdownMenu } from "./DropdownMenu";
import { MemoryRouter } from "react-router-dom";
import { StyledMenu, StyledList } from "../MainNav/MainNav.styled";
import {
  StyledItem,
  StyledTrigger,
  StyledCaret,
  StyledContent,
} from "../MainNav/MainNavItem/MainNavItem.styled";
import { Box } from "@shared/components/Box";
import { GridContainer } from "@shared/components/Grid";

export default {
  title: "Shared/Menus/Dropdown Menu",
  component: DropdownMenu,
  parameters: {
    viewMode: "story",
    previewTabs: {
      "storybook/docs/panel": { hidden: true },
    },
  },
} as ComponentMeta<typeof DropdownMenu>;

const DropdownMenuTemplate: ComponentStory<typeof DropdownMenu> = (
  args: DropdownMenuProps
) => {
  return (
    <MemoryRouter initialEntries={[{ pathname: "/example" }]}>
      <GridContainer>
        <Box
          d="flex"
          justify="space-between"
          align="center"
          py={{ xs: 5, lg: 15 }}
          w="100%"
        >
          <Box d="flex">
            <Box ml="auto" mr={{ xs: 20, md: 40 }}>
              <StyledMenu>
                <StyledList $colGapLarge={40}>
                  <StyledItem>
                    <StyledTrigger>
                      Open Dropdown Nav
                      <StyledCaret icon="ChevronYellow" />
                    </StyledTrigger>
                    <StyledContent>
                      <DropdownMenu {...args} />
                    </StyledContent>
                  </StyledItem>
                </StyledList>
              </StyledMenu>
            </Box>
          </Box>
        </Box>
      </GridContainer>
    </MemoryRouter>
  );
};

export const Default = DropdownMenuTemplate.bind({});
Default.args = {
  subList: LIST_WITH_MIXED_ITEMS,
};
