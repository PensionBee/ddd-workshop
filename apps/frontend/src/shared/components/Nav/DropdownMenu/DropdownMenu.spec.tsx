import { routerRender, vi } from "@shared/testUtils";
import { DropdownMenu } from "./DropdownMenu";
import { LIST, LIST_WITH_MIXED_ITEMS } from "../Nav.content";

describe("DropdownMenu", () => {
  it("should render correctly", () => {
    const { queryByText, asFragment } = routerRender(
      <DropdownMenu subList={LIST} />
    );

    expect(queryByText(LIST[0].children)).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should not render hidden sub list items", () => {
    const { queryByText, asFragment } = routerRender(
      <DropdownMenu subList={LIST_WITH_MIXED_ITEMS} />
    );

    LIST_WITH_MIXED_ITEMS.forEach((item) => {
      if (item.hidden) {
        expect(queryByText(item.children)).toBe(null);
      } else {
        expect(queryByText(item.children)).toBeVisible();
      }
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
