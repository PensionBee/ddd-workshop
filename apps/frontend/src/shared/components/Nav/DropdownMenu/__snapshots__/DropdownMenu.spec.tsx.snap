// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu should not render hidden sub list items 1`] = `
<DocumentFragment>
  <ul
    class="sc-bcXHqe cIfdxZ"
    data-rc="DropdownMenu"
    data-state="closed"
  >
    <li
      class="sc-hLBbgP JISOz"
      data-highlighted="true"
      data-rc="DropdownMenuItem"
    >
      <a
        class="sc-eDvSVe rDWXz"
        href="/"
        tabindex="0"
      >
        <div
          class="sc-dkrFOg hTlDyj"
        >
          Visible Item
        </div>
      </a>
    </li>
    <li
      class="sc-hLBbgP bjcMb"
      data-rc="DropdownMenuItemWithSubList"
    >
      <div
        class="sc-jSUZER"
      >
        <button
          class="sc-gKPRtg fWpcux"
          tabindex="0"
        >
          SubList Item
          <span
            aria-hidden="true"
            class="sc-gswNZR jBXcUP"
            data-rc="LargeSystemIcon.ArrowRight"
          >
            <svg>
              arrow-right.svg
            </svg>
          </span>
        </button>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;

exports[`DropdownMenu should render correctly 1`] = `
<DocumentFragment>
  <ul
    class="sc-bcXHqe cIfdxZ"
    data-rc="DropdownMenu"
    data-state="closed"
  >
    <li
      class="sc-hLBbgP JISOz"
      data-highlighted="true"
      data-rc="DropdownMenuItem"
    >
      <a
        class="sc-eDvSVe rDWXz"
        href="/"
        tabindex="0"
      >
        <div
          class="sc-dkrFOg hTlDyj"
        >
          Visible Item
        </div>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
